PINSFV Incompressible porous flow finite volume capability

GiudGiud opened the pull request:
Address #16756 

Porosity jumps are still WIP, see #17087 

TODO:
- [x] Documentation
- [x] Investigate energy diffusion term, why is there porosity?
- [x] Find the right level of detail for the friction kernel. Useful to pronghorn or simple?

GiudGiud committed:
Add convective heat transfer term, refs #16756

GiudGiud committed:
Add energy time derivatice porous flow kernel

GiudGiud committed:
Add PINSFV Energy advection kernels

GiudGiud committed:
Add porous flow INSFV energy diffusion kernel

GiudGiud committed:
Add porous media INSFV mass equation kernels

GiudGiud committed:
Add porous media momentum time derivative, simplify some other terms through inheritance

GiudGiud committed:
Add PINSFVVelocityVariable to force superficial velocity type
Add momentum advection kernels

GiudGiud committed:
Add momentum diffusion with spatially constant porosity

GiudGiud committed:
Add porosity gravity and boussineq kernels

GiudGiud committed:
Rename gravity and boussinesq kernels for their respective variable
Add momentum friction kernel

GiudGiud committed:
Add momentum pressure gradient term kernel, refs #16756

GiudGiud committed:
Add a basic channel test

GiudGiud committed:
Got the porosity face value approach for momentum advection

GiudGiud committed:
Add documentation to FV Dirichlet boundary conditions

GiudGiud committed:
Fix momentum diffusion term, only enable porosity gradient term for smooth porosity
Interpolate k_eps on face for energy diffusion

GiudGiud committed:
Remove DOFmap printing from INSFV tests

GiudGiud committed:
Add ifdef against local AD indexing, first stab at pinsfv tests

GiudGiud committed:
Move files to match conventional moose file locations

GiudGiud committed:
Rename Gravity and Boussinesq force kernels to highlight that they apply to the momentum equation

GiudGiud committed:
Interpolate porosity with the other terms for momentum advection

GiudGiud committed:
Add symmetry BCs, rename heat convection, add heated test

GiudGiud committed:
Add PINSFV conservation test

GiudGiud committed:
Add porosity to Rhie Chow interpolation, pressure and diffusion terms

GiudGiud committed:
Add more options to PINSFV friction models

GiudGiud committed:
Remove porosity derivative term

GiudGiud committed:
Fix some porous kernels

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/688704/) on e9a4fea wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/17088/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/17088/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 7b30e5ab9fa379e5a73bfc3c375e3fac56251206`

GiudGiud committed:
Add effective thermal diffusion kernel

GiudGiud committed:
Improve params documentation
Compute porosity on the face for symmetry BCs

GiudGiud committed:
Add stubs

GiudGiud committed:
Add discontinuous porosity test

lindsayad commented:
I see that you're not currently using `PINSFVMomentumFriction` in any inputs. I was curious about this because I was thinking about whether it would make sense to migrate Pronghorn drag-computation materials over to the module. However, the material names feel pretty NE-related, e.g. `PebbleBedDragCoefficients` and `KTADragCoefficients`. What are your thoughts on this? The most recent thoughts from DOE leadership is that we might be able to open source mathematics but not realistic materials/geometries.

GiudGiud commented:
Agreed. Based on offline conversation, I ll keep the kernels in and test them with a generic anisotropic friction coefficient.

GiudGiud committed:
Add a flux version of the momentum equation pressure term

GiudGiud committed:
Add a porosity discontinuity term which causes a proportional pressure jump

GiudGiud committed:
Add a Generic vector material for testing anisotropic friction kernels

lindsayad commented on modules/navier_stokes/include/fvkernels/PINSFVMomentumFriction.h:
```diff
@@ -0,0 +1,41 @@
+/**
+ * Imposes a friction force on the momentum equation in porous media.
+ */
+class PINSFVMomentumFriction : public FVElementalKernel
```
For some objects, such as this one, I expect there to be no difference between the PCNSFV and PINSFV implementations. In my rebase on your work, I have been renaming objects such as these from `PINSFV\(.*\)` to `PNSFV\1`. I'm fine continuing to do that work myself, but you may also think about doing it upstream in this PR. So far I believe this object, as well as `PNSFVMomentumPressure` (the flux version) and `PNSFVMomentumGravity` are the objects I've renamed

lindsayad replied:
I understand there may be some confusion for users if some objects have `PINSFV` prefix and others have `PNSFV` prefix. In that case, I'd suggest maybe we do in the .h file:

```
using PINSFVMomentumFriction = PNSFVMomentumFriction;
```

and in the .C file:
```
registerMooseObject("NavierStokesApp", PINSFVMomentumFriction);
registerMooseObject("NavierStokesApp", PNSFVMomentumFriction);
```

GiudGiud replied:
ok I ll let you do the renaming

GiudGiud replied:
We could do that or we could assume users are going to use actions anyway.
How does this do for documentation? Will we have to have two entries and two pages?

lindsayad replied:
Yea I think the assumption about actions is fine. I actually have no idea how documentation would work for the way I proposed

GiudGiud replied:
For a number of INSFV objects that do not change in PINSFV, I have not duplicated them. For example INSFVNaturalFreeSlipBC. I guess we should decide if we want to register PINSFV for everything now or just allow the mixing of PINSFV, INSFV, NSFV depending on what is valid when.

lindsayad commented on modules/navier_stokes/src/fvkernels/PINSFVMomentumFriction.C:
```diff
@@ -0,0 +1,72 @@
+ADReal
+PINSFVMomentumFriction::computeQpResidual()
+{
+  ADReal friction_term;
```
I just got bit by this. This is default initialization with the data left in indeterminate states. You need to explicitly initialize this, e.g.:
```c++
ADReal friction_term = 0;
```

lindsayad commented on modules/navier_stokes/src/fvkernels/PINSFVMomentumFriction.C:
```diff
@@ -0,0 +1,72 @@
+{
+  ADReal friction_term;
+  if (_use_linear_friction_matprop)
+    friction_term += (*_linear_friction_matprop)[_qp] * _u[_qp];
+  if (_use_quadratic_friction_matprop)
+    friction_term += (*_quadratic_friction_matprop)[_qp] * _u[_qp] * std::abs(_u[_qp]);
+  if (_use_Darcy_friction_model)
+    friction_term += (*_cL)[_qp](_component) * _u[_qp];
+  if (_use_Forchheimer_friction_model)
+    friction_term += (*_cQ)[_qp](_component) * _u[_qp];
```
If the user is using `superficial_rho_u` etc. as their nonlinear variable, then these residuals won't be quite right. The `VarMaterial` way would be to retrieve a `momentum` material property and index that by component

lindsayad replied:
Oopsie I’m thinking from a compressible mindset when your object name is still PINSFV. I think u should still change to momentum though right?

lindsayad replied:
Hmm I see that the GenFoam paper uses a formulation proportional to velocity, whereas a lot of Pronghorn publications posit a term proportional to momentum. So I suppose you're probably fine here. Or maybe make an optional material property parameter that if supplied is used in the residual, and if not supplied then `_u` is used?

GiudGiud replied:
ok let's use momentum to be consistent with the closures in Pronghorn for W.

GiudGiud committed:
Add 1D porosity jump test, clean up 2D test
Add porosity jump treatment to advection kernel
Add continuous porosity gradient term from advection kernel
Add effective viscosity option to momentum diffusion
Add momentum matprop option to friction kernel, switch to momentum for consistency with Pronghorn
Fix formulation of porosity gradient term

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/699252/) on a752787 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/17088/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/17088/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format af03b2bda46f1640ae4179d0624be4b0ef0578f7`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/699350/) on e7644df wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/17088/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/17088/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format af03b2bda46f1640ae4179d0624be4b0ef0578f7`

GiudGiud committed:
Update documentation
Express friction in terms of superficial velocity
Remove mass time derivative kernel
Remove option for effective viscosity in momentum diffusion kernel

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/699399/) on 0076253 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/17088/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/17088/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format af03b2bda46f1640ae4179d0624be4b0ef0578f7`

GiudGiud committed:
Clang formatting

GiudGiud committed:
Clang formatting
Fix eps bug in friction

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/699423/) on c91c494 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/17088/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/699567/) on dd72921 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/17088/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/17088/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 62bbfbbb5aa17a8822ac5801e6dafd3529ffcbf1`

GiudGiud committed:
Documentation fixup, add PINSFV main documentation file, refs #16756

GiudGiud committed:
Switch PINSFVVelocityVariable to SuperficialVelocityVariable

lindsayad commented on framework/src/materials/GenericConstantVectorMaterial.C:
```diff
@@ -0,0 +1,57 @@
+  : Material(parameters),
+    _prop_name(getParam<std::string>("prop_name")),
+    _prop_value(getParam<RealVectorValue>("prop_value")),
+    _property(declareProperty<RealVectorValue>(_prop_name))
```
If you want to do the `is_ad` thing (which I'm glad you are!), you need to do `declareGenericProperty<RealVectorValue, is_ad>` here

lindsayad reviewed (changes requested):
@GiudGiud please let me know when you're ready for a formal review

GiudGiud commented:
Will do. I'm still concerned about oscillations when including porosity gradients.
Also hitting FPEs when computing the norm of the gradient of porosity. I m not sure if just the AD part is the concern here.

lindsayad commented:
> Also hitting FPEs when computing the norm of the gradient of porosity

Are you making porosity an AD property? I would make it a regular property unless it's going to depend on a nonlinear variable (I'd only imagine that would be the case in microstructure simulations). I would imagine that in most cells in most simulations the gradient of porosity will be zero, in which case our norm computation in `TypeVector::norm` would put NaNs in the AD derivative vector.

GiudGiud commented:
Porosity is an AuxVariable, so   const MooseVariableFV<Real> * const _eps_var;

I am getting the FPE when doing adGradSln on this variable

lindsayad commented:
That’s a bit unexpected to me. Probably should get a backtrace and/or hop in the debugger. Probably a pretty easy bug to fix. 

> On Mar 14, 2021, at 11:18 AM, Guillaume Giudicelli ***@***.***> wrote:
> 
> ﻿
> Porosity is an AuxVariable, so const MooseVariableFV * const _eps_var;
> 
> I am getting the FPE when doing adGradSln on this variable
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or unsubscribe.

GiudGiud committed:
Temporary fix on FVBCs for neighbor element evaluations
Fix PINSFVSymmetryVelocityBC issue with porosity value
Clean up of MomentumPressureFlux
Make energy advection use porous RC
Tolerance on floating point math for disabling RC at porosity jumps

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/704457/) on 6ebb26d wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/17088/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/17088/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 62bbfbbb5aa17a8822ac5801e6dafd3529ffcbf1`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/704482/) on 31031ec wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/17088/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/17088/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 62bbfbbb5aa17a8822ac5801e6dafd3529ffcbf1`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/704507/) on cc606b9 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/17088/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/17088/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 62bbfbbb5aa17a8822ac5801e6dafd3529ffcbf1`

lindsayad commented:
I think you need a clang-format save hook for your editor haha

> On Mar 14, 2021, at 7:03 PM, MOOSE Build ***@***.***> wrote:
> 
> ﻿
> Job Precheck on cc606b9 wanted to post the following:
> 
> Your code requires style changes.
> 
> A patch was auto generated and copied here
> You can directly apply the patch by running, in the top level of your repository:
> 
> curl -s https://mooseframework.inl.gov/docs/PRs/17088/style.patch | git apply -v
> 
> Alternatively, with your repository up to date and in the top level of your repository:
> 
> git clang-format 62bbfbbb5aa17a8822ac5801e6dafd3529ffcbf1
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or unsubscribe.

GiudGiud commented:
Probably. 
git-clang format doesnt do anything on  any of my machines for some reason. I must be doing something wrong

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/704534/) on 499d559 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/17088/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/17088/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 62bbfbbb5aa17a8822ac5801e6dafd3529ffcbf1`

GiudGiud committed:
Add check on porous variable type for porous kernels

GiudGiud committed:
Fix some PINSFV documentation
Fix GenericVectorMaterial for AD

GiudGiud committed:
Clean up and update PINSFV tests, add friction and boussinesq tests, refs #16755 #16872

GiudGiud committed:
Experimental treatment of porosity jumps in momentum pressure term

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/711553/) on bdbe22c wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/17088/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/17088/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format c6b5d6631ee854188fcfc7ddb966742d76ea6617`

GiudGiud committed:
Improve headers
Allow user to specify what consistutes a discontinuity in porosity

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/711736/) on c697c19 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/17088/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/17088/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format c6b5d6631ee854188fcfc7ddb966742d76ea6617`

GiudGiud committed:
Add all test gold files for PINSFV
Add some variable type checks
Clean source files, leave TODOs for namespace issue with superficial velocity in continuous porosity case
Add a non-unity porosity in conservation test
Update momentum pressure flux kernel to evaluate eps*P on faces
Use PINSFV outflow kernels instead of INSFV for mean pressure channel
Use LU preconditioner for tests that fail in parallel
Remove triangle mesh test for now

Address Alex's 2 reviews
Fix BC execution on neighbor side
Fix ConstantVectorMaterials
Fix documentation
Re-use rcCoeff from INSFVMomentumAdvection

Fix parallel bug in PINSFV momentum advection
Fix potential for bad material properties in MomentumDiffusion
Add 1D and 3D tests to INSFV

Remove linear and quadratic generic friction model, which are present in INSFV

Re-gold tests
Add test for effective diffusion of energy

Use non-unity porosity in heated tests, re-gold

Fix for sparse AD indexing compile
Clang format

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/711972/) on f39594a wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/17088/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/17088/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format c6b5d6631ee854188fcfc7ddb966742d76ea6617`

GiudGiud commented:
I removed the special treatment of Rhie Chow for porosity discontinuities. We will pursue this as a later effort, and also try interfacekernels which should be a good workaround.

I also removed the continuous porosity gradient term for momentum diffusion. I was getting a segfault on taking the gradient of porosity in the neighbor element, so it was not ready anyway.

I left the porosity change tests in. While there are oscillations, they should still show the right pressure drop, as long as the flux formulation for the momentum pressure term that Alex derived is used.
They also show a satisfying match between the discontinuous jump and the continuous jump. 
The average interpolation results show the expected flat superficial velocity profile, but the checkerboard in pressure appears in some cases (unclear why not all cases). 

@lindsayad this should be ready for a review now

GiudGiud commented:
Looks like PINSFV struggles with threads and with MPI, lots of diffs

lindsayad commented on modules/navier_stokes/src/fvkernels/PINSFVMomentumPressureFlux.C:
```diff
@@ -0,0 +1,64 @@
+ADReal
+PINSFVMomentumPressureFlux::computeQpResidual()
+{
+  ADReal p_interface;
+
+  Moose::FV::interpolate(Moose::FV::InterpMethod::Average,
+                         p_interface,
+                         _p_elem[_qp],
+                         _p_neighbor[_qp],
+                         *_face_info,
+                         true);
+  return _eps[_qp] * p_interface * _normal(_index);
```
At one point in time you interpolated epsilon and pressure together. Now you are not. Why is that? This way is not conservative. The neighbor porosity is not entering into the calculation at all...I'm a little confused as to why the change.

GiudGiud replied:
The term is eps * grad_P or eps * div (**I** P). So I take out (assumed constant) eps from the volume integral before applying the divergence theorem.

To have eps * P evaluated at the faces, we would need div(eps P **I**), which we could make appear with an IBP but then there's a porosity gradient which isnt great either.

GiudGiud replied:
The concern I guess is that depending on which element is chosen when the flux kernel is evaluated on the face, the contribution to the residual will be different. That could explain the difficulties in parallel.

I ll change that

GiudGiud commented:
Diffs between parallel and serial/threads on the Rhie Chow porosity jump tests. Hopefully fixing this is the key to figuring those out.

lindsayad commented:
Are you doing anything major with this PR right now, or should I review it? I know you said it's ready but with the recent force push, wanted to double check...

GiudGiud commented:
There are definitely still some issues with parallelism so it's not ready. 
However since this is a big PR now, I think another review now would be helpful

lindsayad commented on framework/include/materials/GenericConstantVectorMaterial.h:
```diff
@@ -0,0 +1,33 @@
+//* This file is part of the MOOSE framework
```
This is definitely a useful class. Thanks for adding!

lindsayad commented on framework/include/materials/GenericConstantVectorMaterial.h:
```diff
@@ -0,0 +1,33 @@
+  virtual void initQpStatefulProperties() override;
+  virtual void computeQpProperties() override;
+
+  std::string _prop_name;
+  const RealVectorValue & _prop_value;
```
Seems like it might be useful to allow a user to create an arbitrary number of these vector properties like we do for scalar material properties? I could imagine a user trying to use this class the first time as they do with the scalar version and maybe being surprised that they can't

GiudGiud replied:
_prop_values(getParam<std::vector<RealVectorValue>>("prop_values"))
only gets the first component of the vector if I try to define multiple vectors. 
Should I be adding a new template to getParam ?

GiudGiud replied:
I ll just use a vector of Real and keep track of indices.

lindsayad replied:
Wait `getParam<std::vector<RealVectorValue>>()` doesn't return a `std::vector<RealVectorValue>`? If there's something wonky there, that's a bug we definitely need to fix. We shouldn't have to do any work-arounds.

GiudGiud replied:
I think the issue lies with RealVectorValue. I was looking into why I need to regold. And my guess is that getParam<RealVectorValue>() already doesnt work

lindsayad replied:
That would be pretty crazy. We do `RealVectorValue` for things like gravity in Navier-Stokes and I've never noticed an issue. Seems like we would have gotten user reports if that didn't work. I could believe that `std::vector<RealVectorValue>` isn't working though because I don't think we that anywhere in the repository yet

lindsayad replied:
I don't want to make you spend a lot of time on this. You've got other things to do and what really matters is the PINSFV stuff. I'm totally fine if you decide to revert to the single prop design. We can add vector ability later in follow-on work if we want.

lindsayad replied:
Actually what you're doing with the vector of `Real` stuff would be a great stand-in if it works. Because the input-file syntax would not change if/when we do switch over to `std::vector<RealVectorValue>`

GiudGiud replied:
Ok I ll post an issue and leave it at that

lindsayad commented on framework/src/fvbcs/FVFluxBC.C:
```diff
@@ -174,7 +174,8 @@ FVFluxBC::computeJacobian(const FaceInfo & fi)
   };
 
-  _assembly.processDerivatives(r, _var.dofIndices()[0], _matrix_tags, local_functor);
+  if (_var.dofIndices().size() > 0)
+    _assembly.processDerivatives(r, _var.dofIndices()[0], _matrix_tags, local_functor);
```
I think I encountered this recently. If we don't have any dof indices here, then we are in big trouble. It means we're not applying a BC when we should be. What I suspect is true is that the dof indices live on the neighbor side of the face. In this case we should either be using data from the neighbor side and feeding the result into the "neighbor" residual (e.g. the row corresponding to `_var.dofIndicesNeighbor()[0]`) or we should be erroring out here. We shouldn't let this stand!

GiudGiud replied:
I think that is the correct fix.
Thanks!

lindsayad reviewed (changes requested):
I'll be reviewing in fits and spurts. Got through the framework directory :smile:

GiudGiud commented on framework/include/materials/DerivativeMaterialInterface.h:
```diff
@@ -49,7 +49,7 @@ class DerivativeMaterialInterface : public T, public DerivativeMaterialPropertyN
   /**
    * Methods for declaring derivative material properties
-   * @tparam U The material property type
+   * @param U The material property type
```
my bad I thought it was a typo

lindsayad commented on modules/navier_stokes/doc/content/modules/navier_stokes/pinsfv.md:
```diff
@@ -0,0 +1,146 @@
+
+!listing modules/navier_stokes/test/tests/finite_volume/pins/channel-flow/2d-rc.i block=FVKernels/u_viscosity
+
+The pressure gradient term in the momentum equation is expressed using a `PINSFVMomentumPressure` term.
```
Should we recommend the `Flux` version instead? I like that one :smile:

GiudGiud replied:
It may be slower ^^

lindsayad replied:
Unless it converges better...then it could be faster.

lindsayad commented on modules/navier_stokes/doc/content/modules/navier_stokes/pinsfv.md:
```diff
@@ -0,0 +1,146 @@
+Both Rhie Chow interpolation and the momentum equation feature a pressure gradient, and the superficial momentum
+advection and diffusion terms feature a porosity or inverse porosity gradient. These are ill-defined near the
+discontinuity. To deal with this issue, the discontinuous case uses a flux-based formulation of the pressure gradient,
+`PINSFVMomentumPressureFlux` and Rhie Chow interpolation reverts to average interpolation on the faces near the discontinuity.
```
Does the Rhie Chow interpolation still revert to average interpolation near the discontinuity?

GiudGiud replied:
I took out that in the latest commit.

GiudGiud replied:
so oscillations in the no-pressure-gradient cases are back

lindsayad replied:
Got it. Yea guessing we'll just have to `FVInterfaceKernel` if we really want to suppress that.

lindsayad reviewed (changes requested):
I'm interested most in how we're doing with the jump tests. It would be nice to see some of those plots that you've shared in slack reproduced here on this PR for both review and for posterity.

GiudGiud commented on modules/navier_stokes/src/fvkernels/PINSFVMomentumAdvection.C:
```diff
@@ -0,0 +1,393 @@
+}
+
+PINSFVMomentumAdvection::PINSFVMomentumAdvection(const InputParameters & params)
+  : INSFVMomentumAdvection(params),
```
sure I can reuse rcCoeff, good call

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/720156/) on c572db1 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/17088/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/17088/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 8b7394a261d838f23abc6744bf4cb16707c8ecf2`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/727184/) on a70352c wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/17088/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/17088/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 06bc1f73f0d4dfb3d01ce9ba77d6bb97346e07eb`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/728231/) on ff07d69 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/17088/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/17088/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 06bc1f73f0d4dfb3d01ce9ba77d6bb97346e07eb`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/728363/) on 0e358f3 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/17088/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/17088/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 06bc1f73f0d4dfb3d01ce9ba77d6bb97346e07eb`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/728391/) on f996c40 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/17088/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/17088/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 06bc1f73f0d4dfb3d01ce9ba77d6bb97346e07eb`

GiudGiud commented:
If I can't figure out a good RC formulation for porosity jumps by the end of the review, let's just delete the tests for it?

lindsayad commented:
> I'm enclined to keep them but the one that's inconsistent between serial and parallel is kind of problematic.,

You can specify `max_parallel = 1` in your `tests` spec if you want

lindsayad commented on modules/navier_stokes/include/fvkernels/PINSFVEnergyTimeDerivative.h:
```diff
@@ -0,0 +1,31 @@
+
+#include "FVTimeKernel.h"
+
+class PINSFVEnergyTimeDerivative : public FVTimeKernel
```
doxygen please

GiudGiud replied:
ok adding them to all fvbcs

lindsayad replied:
I think this might be a hint that I did not properly doxygen all the INSFV classes I added. If so then touche :smile:

lindsayad commented on modules/navier_stokes/include/fvkernels/PINSFVMomentumDiffusion.h:
```diff
@@ -0,0 +1,35 @@
+#include "FVFluxKernel.h"
+
+/**
+ * A flux kernel diffusion of momentum in porous media across cell faces
```
```suggestion
 * A flux kernel for diffusion of momentum in porous media across cell faces
```

lindsayad commented on modules/navier_stokes/include/variables/PINSFVSuperficialVelocityVariable.h:
```diff
@@ -0,0 +1,22 @@
+
+class InputParameters;
+
+class PINSFVSuperficialVelocityVariable : public INSFVVelocityVariable
```
doxygen

lindsayad commented on modules/navier_stokes/src/fvbcs/PINSFVMomentumAdvectionOutflowBC.C:
```diff
@@ -0,0 +1,90 @@
+#endif
+
+  if (!_u_var)
+    paramError("u", "the u velocity must be an PINSFVSuperficialVelocityVariable.");
```
```suggestion
    paramError("u", "the u velocity must be a PINSFVSuperficialVelocityVariable.");
```
and for other similar

lindsayad commented on modules/navier_stokes/src/fvkernels/PINSFVMomentumAdvection.C:
```diff
@@ -0,0 +1,390 @@
+    return;
+
+  // Avoid computing a pressure gradient near porosity jumps
+  if (!_smooth_porosity)
```
I suppose this member just allows for some optimization to potentially avoid calling gradient reconstruction?

GiudGiud replied:
it's not optimization, it's to zero the RC term when there s a porosity jump

lindsayad replied:
No I mean you would get the correct technical behavior without this user provided `_smooth_porosity` parameter right? If you removed `_smooth_porosity` then your direct checking of the porosity gradients would tell you whether you have porosity jumps or not. My concern is that relying on the user provided `smooth_porosity` value can be prone to error. For instance what if they say `smooth_porosity = true` when in fact it isn't? Then you you would go ahead and *not* zero the RC term near a jump.

lindsayad commented on modules/navier_stokes/src/fvkernels/PINSFVMomentumAdvectionPorosityGradient.C:
```diff
@@ -0,0 +1,73 @@
+    mooseError("PINSFVMomentumAdvectionPorosityGradient may only be used with a superficial "
+               "velocity variable, of variable type PINSFVSuperficialVelocityVariable.");
+
+  if (!_smooth_porosity)
+    paramError(
+        "smooth_porosity",
+        "The MomentumAdvectionContinuousPorosity may only be used with a continuous porosity.");
```
Is the intention here to remove this restriction sometime in the future? Currently the default value for `smooth_porosity` will trigger an error for a parameter that the user didn't even set.

GiudGiud replied:
the intent was to force the user to set smooth_porosity.
I guess I ll make it a required param

GiudGiud replied:
If in the future we find a volumetric formulation of this part of the eps * grad_P IBP that does well on porosity jump, we would remove that error for sure.

lindsayad replied:
This still feels a bit wacky to me. We are adding a required boolean parameter that only has one permissible value. But I do understand where you're coming from. You want to make sure the user understands that they should not be using this object if the porosity is not smooth. Ideally we could do a check during object construction (not technically possible) or during `initialSetup` (technically possible but O(n_face) expensive)) of whether the porosity is smooth and then error out of it is not. That would negate the need for this parameter which only has one permissible value. But acknowledging the drawbacks of these automated error checks I think what you've got here is fine

lindsayad commented on modules/navier_stokes/test/tests/finite_volume/pins/channel-flow/tests:
```diff
@@ -0,0 +1,206 @@
+    exodiff = rc-heated.e
+    method = "!dbg"
+    abs_zero = 1e-9
+    requirement = 'The system shall be able to solve for fluid energy diffusion, advection and convection with the local medium in a 2D channel.'
```
"local medium in a 2D channel" isn't clear to me

lindsayad commented on modules/navier_stokes/test/tests/finite_volume/pins/channel-flow/tests:
```diff
@@ -0,0 +1,206 @@
+    exodiff = rc-heated-rz.e
+    method = "!dbg"
+    abs_zero = 1e-9
+    requirement = 'The system shall be able to solve for fluid energy diffusion, advection and convection with the local medium in rz geometry.'
```
Same as above

lindsayad commented on modules/navier_stokes/test/tests/finite_volume/pins/channel-flow/tests:
```diff
@@ -0,0 +1,206 @@
+    exodiff = rc-heated-effective.e
+    method = "!dbg"
+    abs_zero = 1e-9
+    requirement = 'The system shall be able to solve for fluid energy effective diffusion, advection and convection with the local medium in a 2D channel.'
```
Yea just unclear on what local medium is here. Is there a global medium? Is that the surroundings? If the distinction is our simulation domain is local then presumably the 2D channel is local too. I suggest clarifying any requirement in this `tests` spec that currently has `local medium`

GiudGiud replied:
ok I ll replace local medium by solid phase

lindsayad reviewed (changes requested):
General comments

- You shimm'd some of the kernel classes so that you can present `PINSFV*` acronyms for all your physics (at least I think you did it for all) even when all you did was implement `validParams` and the constructor. Do you want to do the same with the boundary conditions? I saw inputs with all PINSFV for kernels and all INSFV for BCs. I honestly don't have a huge opinion on this... it's nice from a user perspective I think to have it all the same but it's annoying from a development stand-point. If eventually we're going to wrap everything in actions, then maybe we don't worry about shimming stuff when we don't actually need to implement anything new
- You have a couple huge test directories. Do you want to break it out into sub-directories with separate focuses? Tests would run faster that way and it may help a user navigate to their particular interest quicker.

Overall I think there's not a lot to do here before merging (keeping in mind that we know our current limitations).

lindsayad commented on modules/navier_stokes/doc/content/source/fvkernels/PINSFVMomentumPressure.md:
```diff
@@ -0,0 +1,14 @@
+
+This object adds the $\epsilon \nabla p$ term of the
+incompressible Navier Stokes momentum equation as a volumetric term.
+This formulation causes oscillation near porosity discontinuities and the
+[PINSFVMomentumPressureFlux.md] kernel should be preferred then.
```
When you run with the flux formulation do you also include the second term that comes out of the integration by parts? E.g. the p * grad_eps? If not, then I wonder if that's the reason that the flux version "performs better"

GiudGiud replied:
Not without friction, or if the friction term is the same on both sides (which it would be)

It's definitely the reason it performs better. We're getting rid of the gradient near a discontinuity by doing this. The pressure or the porosity gradient both don't help

lindsayad replied:
Hmm I feel like the result will be wrong if we totally disregard the gradient (either pressure or porosity)

GiudGiud replied:
I ll run the smooth case again but I think the result was closer without the gradient. That, the genFoam paper getting rid of the advection porosity gradient terms and the lack of oscillations (in some cases) were the main reasons I removed it

GiudGiud replied:
Pressure drop should be correct with this div(eps * P) formulation, since by definition it conserves that term

lindsayad replied:
It definitely conserves eps * P, but that's not the term in the governing equation nor for example does div(eps*P) (by itself) yield the correct physical results for compressible fluids. But it does for incompressible fluids then that would be marvelous

GiudGiud commented:
> General comments
> 
> * You shimm'd some of the kernel classes so that you can present `PINSFV*` acronyms for all your physics (at least I think you did it for all) even when all you did was implement `validParams` and the constructor. Do you want to do the same with the boundary conditions? I saw inputs with all PINSFV for kernels and all INSFV for BCs. I honestly don't have a huge opinion on this... it's nice from a user perspective I think to have it all the same but it's annoying from a development stand-point. If eventually we're going to wrap everything in actions, then maybe we don't worry about shimming stuff when we don't actually need to implement anything new

The only class I did that for was PINSFVMomentumTimeDerivative. It's because I started out with a porosity time derivative. I could delete it
I wasn't planning on doing the same for the FVBCs. 

> * You have a couple huge test directories. Do you want to break it out into sub-directories with separate focuses? Tests would run faster that way and it may help a user navigate to their particular interest quicker.

Ok I ll break the heated tests away. 

> Overall I think there's not a lot to do here before merging (keeping in mind that we know our current limitations).

Wouhou!! The end of the tunnel. Ok I ll address everything tomorrow

GiudGiud committed:
Apply suggestions from code review

Co-authored-by: Alex Lindsay <alexlindsay239@gmail.com>

GiudGiud committed:
Re-organize PINSFV channel tests

GiudGiud commented on framework/doc/content/source/materials/GenericConstantVectorMaterial.md:
```diff
@@ -0,0 +1,36 @@
+# GenericConstantVectorMaterial
+
+!alert! construction title=Undocumented Class
+The GenericConstantVectorMaterial has not been documented. The content listed below should be used as a starting point for
+documenting the class, which includes the typical automatic documentation associated with a
+MooseObject; however, what is contained is ultimately determined by what is necessary to make the
+documentation clear for users.
```
but all the other generic materials dont have anything haha

I'll add that

GiudGiud committed:
Add documentation to generic materials, refs #17566

GiudGiud committed:
Fix up documentation
Address Alex's review:
- add docstring to fvbcs
- fix equation in momentum outflow bc
- use solid phase instead of local medium in test docs
- break up pinsfv channel test directory

GiudGiud commented:
grad(eps*P) term
![image](https://user-images.githubusercontent.com/25037151/116624877-f4af6c80-a905-11eb-867a-f6a35eb4a49f.png)
eps * grad_P
![image](https://user-images.githubusercontent.com/25037151/116625066-42c47000-a906-11eb-8db7-7df05ced1073.png)
grad(eps*P) + P * grad_eps
![image](https://user-images.githubusercontent.com/25037151/116625008-27596500-a906-11eb-8db7-731371eff0a5.png)

the gradient terms are definitely part of the problem.
Notice that these are with the 'zeroing the RC term' in the advection kernel

lindsayad commented:
> The only class I did that for was PINSFVMomentumTimeDerivative. It's because I started out with a porosity time derivative. I could delete it

Ugh this is tough. I already have "dumb" shims for `INSFVEnergyVariable` and `INSFVScalarFieldVariable`. Ok I'll give you a choice:

- Remove the `PINSFVMomentumTimeDerivative` shim and also remove the shims that are already there for the variables
- Don't remove your shim or the existing ones

If you elect the latter, then I still think we should erase the shims in a later PR. It just increases maintenance burden to have more files floating around (recognizing that I am the bad actor currently)

lindsayad commented:
I wish you could thread on github like you can on gitlab. Anyway, re: different eps * grad_p implementations:

Wow! These answers are all so different! I'm glad you made this plot. Ok I assume that "u" here is the superficial velocity?

lindsayad commented:
Also what is the density?

GiudGiud commented:
Density is 1.1
I was thinking about this and I need to redo the plot with the two terms from the IBP. I used the advection porosity gradient term (rho V V grad_eps / eps^2), not the pressure gradient porosity gradient term (P grad eps)
my bad. I ll update the post and ping when it's done.

Ok I'll leave the shims. They are not a big deal for maintenance right now, they're very easy to maintain since they do nothing special.

lindsayad commented:
Oh also what is porosity change? 0.5 to 1 or something like that?

GiudGiud commented:
Jump from 1 to 0.5

GiudGiud commented:
Updated the figure. They are still all different results.
While the oscillations does depend on the mesh, it doesn't get worse for option 3 with refinement, and the pressure drop converges.

lindsayad commented:
Using the convention of assigning the value at a step change to be the average of its value on either side of the discontinuity, I compute that p on the LHS of the domain should be two and seven fifteenths (or 2.4667). My beautiful rotated hand-work is attached. By that scratch work it looks like options 2 and 3, though oscillatory, are much closer to the truth.

![IMG_1075](https://user-images.githubusercontent.com/10248304/116912016-7cde8c00-abfc-11eb-8bce-0fbdcc520900.jpg)

GiudGiud commented:
Well here we upwind the superficial velocity and average the 1/eps term (it was an accommodation I made to improve results, I have not found that elsewhere).
So that gives us:
p1 = 1 + 1/eps_face * rho * ( -1 + (eps u) * (eps * u)_upwind * (1/eps)_face) = 1 + 1/0.75*1.1*(-1+1*1*1.5) = 1.733

so in the framework of what I currently have in the advection kernel, option 1 is better.
We could remove what I did in the advection kernel though. It does not have a strong theoretical basis.

lindsayad commented:
I think what we should aim for is the most physical result. What do your results look like if you change from a step change to a middle region of continuous change in porosity?

GiudGiud commented:
I was comparing to the continuous jump case. I was satisfied by the match between continuous and discontinuous. I ll redo it for the 3 options to see if there is some insight as to what expression of the pressure term works better here (with this advection term)

lindsayad commented:
Instead of doing math that may be wrong, I should instead look at [Bernoulli's equation](https://www.khanacademy.org/science/physics/fluids/fluid-dynamics/a/what-is-bernoullis-equation) and using that the pressure on the LHS should be 2.65

lindsayad commented:
I think you said you were solving inviscid equations for those plots?

GiudGiud commented:
No these have viscosity. Jieun is doing inviscid flow. 
These do not have a pressure gradient though, as there are free-slip boundary conditions / it's 1D

lindsayad commented:
Well this is something I've been wrestling with today. Here is a plot with the diverging cylindrical channel INSFV free-slip test adapted to be 2D instead of 2DRZ. I was hoping to benchmark the porosity results against the results of the free-flow problem with varying area. However, the results shown here bring to light the fact that Bernoulli's principle which is derived from conservation of energy is only valid for an inviscid fluid. Viscosity will (only now that I've wrestled with it all day does this seem (more) obvious in hindsight) convert potential (represented by static pressure) and kinetic energy into heat energy such that any attempt to compute the pressure profile based on Bernoulli is invalid/ill-fated.

More viscous has a viscosity of 1. Less viscous has a viscosity of 1e-1. Difference is inversion of the pressure profile.

![Screenshot from 2021-05-03 16-01-10](https://user-images.githubusercontent.com/10248304/116944185-d9a46b80-ac29-11eb-83c3-ae82d79f5cb7.png)

lindsayad commented:
So in the presence of viscosity my attempt to verify your work with my hand-calculations is simply wrong.

Would you be open to doing an MMS test with continuously varying porosity? I would just love to have some concrete result that verifies one or more of the pressure implementations (along with other kernels). Right now I feel a little lost...it seems like we are trying to verify the discontinuous results with the continuous results when we haven't verified the continuous results.

lindsayad commented:
> Well here we upwind the superficial velocity and average the 1/eps term (it was an accommodation I made to improve results, I have not found that elsewhere).
> So that gives us:
> p1 = 1 + 1/eps_face * rho * ( -1 + (eps u) * (eps * u)_upwind * (1/eps)_face) = 1 + 1/0.75 * 1.1 * (-1+1 * 1 * 1.5) = 1.733
> 
> so in the framework of what I currently have in the advection kernel, option 1 is better.
> We could remove what I did in the advection kernel though. It does not have a strong theoretical basis.
> 
> A few other options for the advection kernel to show that the answer varies wildly:
> p1 = 1 + 1/eps_face * rho * ( -1 + (eps u) * (eps * u)_upwind * (1/eps)_upwind) = 1 + 1/0.75 * 1.1 * (-1+1 * 1 * 1) = 1
> p1 = 1 + 1/eps_face * rho * ( -1 + (eps u) * (u)_upwind) = 1 + 1/0.75 * 1.1 * (-1+1 * 1) = 1
> p1 = 1 + 1/eps_face * rho * ( -1 + (eps u) * (eps * u)_upwind * (1/eps_face)) = 1 + 1/0.75 * 1.1 * (-1+1 * 1 * 1.3333) = 1.4888

Like this passage full of different results doesn't give me much confidence. At least we know what PDEs we are trying to solve and with an MMS test we could verify (within at least some subset of conditions) that we converge the approximate solution to an expected analytical solution.

GiudGiud commented:
I think MMS is welcome here to add confidence in our continuous porosity jump solution. I have not done that before. I'll have to find a case for which I know the analytical solution right? What process did you follow for INSFV?

This passage shows the issue with dealing with a jump. There are just lots of options in terms of discretizing it.  
The literature does not agree on anything as far as I've read.

lindsayad commented:
> I'll have to find a case for which I know the analytical solution right?

That's the nice thing about the Method of Manufactured Solutions (with emphasis on manufactured). You just pick a solution and then you compute a forcing function that will help you obtain it, e.g. we add artificial sources/sinks to every equation. You can look in `navier_stokes/test/tests/finite_volume/ins/mms/channel-flow` for some examples. In that directory I compute 1D manufactured solutions/forcing functions with `1d-compute-functions.py` and then the 2D ones with `2d-average-comput-functions.py`. Ignore the name there... it's used for both average and RC.

lindsayad commented:
The MMS testing is then done through `PythonUnitTests` in the `tests` spec which calls to the script `test.py`

lindsayad commented:
Have you had much luck with inviscid(ish) flows with RC? When I was testing the diverging channel free flow problem yesterday I was able to run with 6 orders of magnitude lower viscosity with average velocity interpolation compared to RC.

GiudGiud commented:
Ok I ll look into that
The main other thing to run with low viscosity is to ramp it down through a (pseudo) transient. I've been able to use the real fluid viscosity for the FHR core.

lindsayad commented:
> Ok I ll look into that

To be clear I was testing with INSFV. So this is not your "fault" :smile: I was just curious about your experience. Ok yea ramping is definitely a good strategy. It's interesting that RC is so much worse than average with respect to convergence for low viscosity (at least on the diverging channel).

GiudGiud commented:
![image](https://user-images.githubusercontent.com/25037151/117200382-b227dd00-ada8-11eb-9918-36d6d805ff27.png)
with continuously varying porosity

I ll upload the new tests, update the jumps tests and rebase now

GiudGiud committed:
Add MMS testing of flat and continuously varying porosity
Re-add momentum diffusion porosity gradient term
Interpolate 1/eps term alongside the velocity (so intersititial velocity instead of superficial)
Re-gold porosity jump terms

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/735951/) on 760eedf wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/17088/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/17088/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format af4438b15cc9bb192f34c09b6781203a9939f510`
